[APMON-193] Add support for Single Step Instrumentation for latest library versions

liliyadd opened the pull request:
Previous [PR1](https://github.com/DataDog/datadog-agent/pull/19504)

### What does this PR do?
1. Inject client APM libraries as an init containers if Single Step Instrumentation is enabled. 
2. Sets default configuration to be injected as env variables. The configuration defined by users, or injected as unified service tags takes precedence over default basic configuration. As part of basic configuration DD_SERVICE might be set to the name of the Kubernetes resource that the pod belongs to. Supporting StatefulSet, Job, CronJob, DaemonSet, ReplicaSet.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
https://docs.google.com/document/d/12iU1nU50cGs27H5s2fUmUbxsjSpU2RO4HhAF1got2ds/edit
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

liliyadd committed:
Add configurations for Single Step Instrumentation in k8s

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Move out the logic for checking if APM instrumentation is enabled

liliyadd committed:
Cleanup

liliyadd committed:
Fix spelling

liliyadd committed:
Add skeleton for Single Step Instrumentation

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Address comments and disable SSI in kube-system

liliyadd committed:
Remove unused shouldInjectConfig() and shouldInjectTags()

liliyadd committed:
Fix comment

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Add function to check if pod can be mutated

liliyadd committed:
Change log line

liliyadd committed:
Delete unused function to pass repo tests

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Merge remote-tracking branch 'origin/liliya.belaus/single-step-apm-1' into liliya.belaus/single-step-apm-2

liliyadd committed:
Add shouldMutatePod function

liliyadd committed:
Add single step configurations to config_template

vboulineau commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -427,6 +452,43 @@ func injectLibRequirements(pod *corev1.Pod, ctrName string, envVars []envVar) er
+		return owner.Name, nil
+	case "ReplicaSet":
+		// Remove the suffix in ReplicaSet name to get Deployment name
+		return owner.Name[:strings.LastIndex(owner.Name, "-")], nil
```
We do have a function to do this in `pkg/util/kubernetes/helpers.go`

liliyadd committed:
Fix typos

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-1

liliyadd committed:
Merge remote-tracking branch 'origin/liliya.belaus/single-step-apm-1' into liliya.belaus/single-step-apm-2

liliyadd committed:
Use helper function to get Deployment name from ReplicaSet

Kyle-Verhoog commented on pkg/clusteragent/admission/controllers/webhook/common.go:
```diff
@@ -18,7 +18,9 @@ import (
 	var labelSelector metav1.LabelSelector
 
-	if config.Datadog.GetBool("admission_controller.mutate_unlabelled") {
+	if config.Datadog.GetBool("admission_controller.mutate_unlabelled") ||
```
This will now subtly permit anything else using `mutate_unlabelled` to run if APM is enabled. I did a quick check and it looks like only APM functionality is using `buildLabelSelectors`. 

Since this function is in `common` I think we should rename the function to `buildAPMLabelSelectors` or similar so that if other functionality added in the future tries to use `mutate_unlabelled` it doesn't mistakenly get enabled if APM instrumentation is.

liliyadd replied:
This logic is just limiting a number of pods that Admission Controller will be looking at to potentially mutate them. It's not going to trigger any mutations on its own. So the pods that don't have mutateUnlabelled or APM SSI won't be considered by admission controller unless they have the label set to `true`.

Kyle-Verhoog replied:
oh right 🤦‍♂️ this thing just opts the pods into being filtered out later on. All good here then!

Kyle-Verhoog commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -428,6 +454,42 @@ func injectLibRequirements(pod *corev1.Pod, ctrName string, envVars []envVar) er
+	case "DaemonSet":
+		return owner.Name, nil
+	case "ReplicaSet":
+		return kubernetes.ParseDeploymentForReplicaSet(owner.Name), nil
```
:+1: awesome we have a helper for this!

Kyle-Verhoog commented on pkg/clusteragent/admission/mutate/auto_instrumentation_test.go:
```diff
@@ -964,9 +1075,157 @@ func TestInjectAutoInstrumentation(t *testing.T) {
+				mockConfig.Set("apm_config.instrumentation.enabled", false)
+				mockConfig.Set("apm_config.instrumentation.enabled_namespaces", []string{"ns"})
+			},
+		},
```
I think we should add test cases for the UST tags `tags.datadoghq.com/{env, service, version}` and also make sure that the if the UST service tag is set that we do not override it with the deployment name (user value takes precedence).

liliyadd replied:
We cannot test this logic in this unit-test. Tags injection and library injection are different endpoints of the Admission Controller https://github.com/DataDog/datadog-agent/blob/main/cmd/cluster-agent/subcommands/start/command.go#L327-L329. I can add a test for this use case to the system-tests, but for now tested manually and the UST is respected.

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-2

pgimalac reviewed (approved):
LGTM for ASC file

alai97 reviewed (approved):
Looks good for docs!

liliyadd committed:
Merge remote-tracking branch 'origin/main' into liliya.belaus/single-step-apm-2

sgnn7 reviewed (approved):
LGTM for ASC-owned `pkg/config/config_template.yaml`

liliyadd commented:
I have rebased the next PR in the chain https://github.com/DataDog/datadog-agent/pull/19714 on top of `main` to simplify merge to RC2. With that, all changes in current PR are now added to https://github.com/DataDog/datadog-agent/pull/19714. I'm going to close this PR.
